### PR TITLE
Migrate carto config & basemap modules to Typescript 

### PR DESCRIPTION
**Pull request by zbigg**
For #5589
#### Change List
- Migrate Carto config & basemap modules to Typescript

### CODE REVIEW

File: modules/carto/src/config.ts [hunk 1]
```diff
@@ -4,17 +4,33 @@ export const defaultClassicCredentials = {
...
+}
 
-let credentials = {};
+type Credentials = ClassicCredentials | CloudNativeCredentials;
+let credentials: Credentials = {} as Credentials;
 
 setDefaultCredentials({});
 
-export function setDefaultCredentials({apiVersion, ...rest}) {
+export function setDefaultCredentials({apiVersion, ...rest}: Partial<Credentials>) {
```
Comment (felixpalmer) [inline modules/carto/src/config.ts:33]: Does this have to be Partial? I would instead mark the properties that are optional as optional (that is, everything except `apiVersion`).

File: modules/carto/src/api/maps-api-common.ts [hunk 2]
```diff
@@ -5,37 +5,37 @@ export const API_VERSIONS = {
...
   GEOJSON: 'geojson',
   BINARY: 'binary'
-};
+} as const;
 
 /**
  * Simple encode parameter
  */
-export function encodeParameter(name, value) {
+export function encodeParameter(name: string, value: string) {
```
Comment (felixpalmer) [inline modules/carto/src/api/maps-api-common.ts:39]: Include return value here

**Review by felixpalmer — CHANGES_REQUESTED** [review]
Great job! Just a few minor changes

File: modules/carto/src/config.ts (hunk 1 above)
Comment (zbigg) [reply modules/carto/src/config.ts:33]: In this signature everything is optional, including `apiVersion`. So yes, this is partial.

With those typings, i tried to convey fact that `setDefaultCredentials` can accept almost everything (partial), but `getDefaultCredentials` will return actual "Credentials" for which some of them are actually mandatory and/or filled using defaults.

So, it depends.

With this typings, instantiating `Credentials` is close-to typesafe.

File: modules/carto/src/api/maps-api-common.ts (hunk 2 above)
Comment (zbigg) [reply modules/carto/src/api/maps-api-common.ts:39]: appied

File: modules/carto/src/config.ts (hunk 1 above)
Comment (felixpalmer) [reply modules/carto/src/config.ts:33]: Ideally it shouldn't accept everything. E.g. this mix of parameters is invalid:

```javascript
{
  apiVersion: 'v1',
  accessToken: 'XXXX'
}
```

Would `Partial<ClassicCredentials> | Partial<CloudNativeCredentials>` work?

**Review by felixpalmer — APPROVED** [review]

**Pull request merged**